# <center>MC906/MO416 - Introduction to Artificial Intelligence</center>
# <center>Institute of Computing - Unicamp</center>
# <center>Prof. Esther Colombini</center>

### <center>Eduardo S. Ito (RA159086)</center>
### <center>Lucas Peres  (RA 265193)</center>
### <center>Thales E. Nazatto (RA 074388)</center>

# Conteúdo

1. Introdução
    1. Descrição do problema
2. Metodologia
    1. Modelagem da solução
        1. Requisitos
        2. Implementação
        3. Problema, agentes e ambiente
    2. Algoritmos
        1. Busca não-informada
        2. Busca informada
        3. Busca local (*Hill Climbing*)
3. Resultados
4. Conclusão
5. Referências

# Introdução
**Pac-man** é um jogo para *Arcades* lançado em 1980 pela **Namco**, criado por Toru Iwatani e ilustrado na Figura 1. Nele, o jogador controla uma bolinha amarela (o *Pac-Man* do título) dentro de um labirinto que tem como objetivo comer todas as bolinhas brancas existentes nele enquanto é perseguido por 4 fantasmas de nomes *Blinky*, *Pinky*, *Inky* e *Clyde*. Na época, Iwatani criou o jogo para atrair pessoas de ambos os sexos, uma vez que os gêneros de jogos lançados nessa época, como jogos de guerra e esportes, eram majoritariamente pensados para agradar o público masculino. Tal estratégia se provou um grande sucesso, e *Pac-Man* se tornou uma das mais lucrativas franquias de todos os tempos, gerando mais de US$ 14 bilhões em receita até o ano de 2016 **[1]**.

<img src="https://thumbs.web.sapo.io/?W=775&H=0&delay_optim=1&webp=1&epic=NGRiQ8y+7rDb2cocSbMwl47ynqxTFYLIDm2IhR5IZK5BFQooKY19ghBWe//2n+M5C2MgCiqhexDuDLBr/f/MrofnNA==" height="240" width="320" alt="Tela do jogo Pac-Man" title="Tela do jogo Pac-Man" />
<center><strong>Figura 1.</strong> Tela do jogo <i>Pac-Man</i></center>

Uma dos grandes inovações para a época foi sua Inteligência Artificial. Cada fantasma se comportava de maneira diferente de acordo com os movimentos realizados pelo jogador, que poderiam ter a possibilidade de encurralá-lo. Enquanto *Blinky* perseguia *Pac-Man* diretamente, *Pinky* e *Inky* tentavam se posicionar a sua frente e *Clyde* alternava seus estados entre perseguir o *Pac-Man* e fugir dele. Tais comportamentos fizeram com que *Pac-Man* fosse um caso de estudos no ramo da Inteligência Artificial.



Neste relatório, será mostrada uma modelagem do problema determinado em **[2]** utilizando 5 algoritmos de busca no contexto deste jogo, sendo 2 de busca não-informada (*Breadth-First Search* e *Depth-First Search*), 2 de busca informada (*Greedy Best-First Search* e *A*\*) e 1 de busca local (*Hill Climbing*). Primeiro será detalhada como foi feita a modelagem do problema e as soluções aplicadas. Depois, serão feitos experimentos e dicussões para avaliar o comportamento dos algoritmos e, no final, as conclusões destes experimentos.

## Descrição do problema

Conforme determinado em **[2]**, o problema consiste em implementar 5 algoritmos diferentes de busca, sendo 2 de busca não-informada, 2 de busca informada (com 2 heurísticas diferentes) e 1 de busca local usando o *Pac-man* como tema, tendo como objetivo comer todas as bolinhas presentes no labirinto, caso consiga evitar os fantasmas. Dentre esses, é necessário avaliar critérios como otimalidade, completude e custo computacional para definir qual a melhor solução dentre elas.

O labirinto é representado por um grid, também ilustrado na Figura 2, cujo tamanho e forma devem ser definidos. Deve ser definido também, além das posições das bolinhas que o *Pac-man* deve comer durante a execução do algoritmo, as posições de 3 fantasmas (que serão estáticos) e uma posição final de parada. É necessário especificar também a representação dos estados, as ações realizadas, o teste realizado para se chegar ao objetivo, o custo **g(x)** do caminho percorrido e as heurísticas utilizadas.

<img src="https://shaunlebron.github.io/pacman-mazegen/img/origmaps_2x_print.png" height="360" width="480" alt="Tela do jogo Pac-Man" title="Tela do jogo Pac-Man" />
<center><strong>Figura 2.</strong> Exemplos de labirintos semelhantes <strong>[6]</strong></center>

# Metodologia

## Modelagem da solução

### Requisitos

Dada a descrição do problema, foram levantados os seguintes requisitos para a conclusão deste projeto:

- **[R1]** 2 algoritmos de busca não-informada
- **[R2]** 2 algoritmos de busca informada (com 2 heurísticas diferentes)
- **[R3]** 1 algoritmo de busca local
- **[R4]** Definir como o problema é modelado
- **[R5]** Especificações e restrições da aplicação
- **[R6]** Avaliar custo computacional dos algoritmos
- **[R7]** Avaliar completude dos algoritmos 
- **[R8]** Avaliar otimalidade dos algoritmos 
- **[R9]** Definir o tamanho e formato do labirinto
- **[R10]** Definir uma posição final de parada
- **[R11]** Definir as posições de 3 fantasmas (que serão estáticos)
- **[R12]** Definir a posição inicial do *Pac-man*
- **[R13]** Especificar representação dos estados
- **[R14]** Especificar as ações realizadas
- **[R15]** Especificar o teste realizado para se chegar ao objetivo
- **[R16]** Especificar o custo **g(x)** do caminho percorrido
- **[R17]** Especificar as heurísticas utilizadas
- **[R18]** A posição inicial do *Pac-man* pode variar de acordo com o problema
- **[R19]** A posição final de parada pode variar de acordo com o problema
- **[R20]** Discretização dos estados (tamanho e configuração do labirinto)
- **[R21]** As posições dos fantasmas podem variar de acordo com o problema
- **[R22]** A busca não será *online*. O problema é completamente observável e determinístico

### Implementação

Para a implementação, foi escolhido em vez da biblioteca AIMA **[3]** o código da Universidade de Berkeley **[4]**. Inicialmente a principal motivação de fazer a troca reside no fato que o código já possui as funções gráficas para exibição do problema, fazendo com que o foco das mudanças seja apenas na modelagem e nos algoritmos. Outras vantagens notadas é que a adaptação deste código para a representação de estados e ações foi mais simples e ele já possuía ferramentas para testes do objetivo. Entretanto, após informações futuras foi comprovado que não seria estritamente necessário e, como o Python 3 foi escolhido para a implementação, foram necessárias adaptações devido a diferenças na biblioteca *tkint* **[5]**. 

Originalmente, o algoritmo de Berkeley já tem um agente pacman, um agente de busca (searchAgent), um agente de teclado (keyboardAgent) e agente(s) fantasma(s) (ghostAgent), que pode ter várias réplicas, comida(s) (food) e um layout de um labirinto (maze). A ideia básica é que um layout pode ser construída de forma flexível com formatos genéricos, sobre os quais se colocam os muros (representados por %), a comida (representado por . em quaisquer posição), o pacman (representado por P, em quaisquer posições), os fantasmas (representados por um ou mais G, em quaisquer posições). O objetivo do pacman é comer todas as comidas do labirinto, mesmo que seja apenas uma unitária, e assim ganhar o jogo, o(s) fantasma(s) vai/vão tentar atacar o pacman de forma dinâmica e assim o conseguir, o pacman perde o jogo. O problema do algoritmo foi modelado de forma que há um estado de início do pacman (posição no maze), um objetivo (ou estado final, ou objetivo, ou goal) - originalmente hardcoded, funcão de sucessão e função de custo. Por comando de linha se determina qual o agente de busca (searchAgent) e o algoritmo de busca (não informado, informado, local) será utilizado, mas não implementa nenhuma heurística e parâmetro para apresentação gráfico no labirinto (i.e.o zoom). O algoritmo tem duas partes principais, a determinação da árvore de estado, que é o caminho que será percorrido pelo pacman ao longo do labirinto, em forma de lista de ações (EAST, WEST, NORTH, SOUTH). O qual será utilizado como entrada na segunda etapa para a operação do pacman na representação gráfica. O pacman pode ter uma lista de ações até o objetivo final, mas o planejamento pode ser interrompido por um agente fantasma a qualquer momento e assim ser comido e perder o jogo. 

Não houve uma mudança significativa na segunda etapa onde a lista de ações é submetida à ação da parte gráfica. Apenas foi introduzido um código para pausar os agentes, ora dinâmico, dos fantasmas. Ainda assim, com capacidade de matar o agente pacman. Foram extendidos a atual capacidade do agente de busca para cumprir com os requisitos do Projeto 1.

A principal mudança realizada em relação ao código original os arquivos **searchAgents.py** and **search.py**, que possuíam os códigos para modelagem dos agentes e para os algoritmos de busca, foram substituídos por versões customizadas, batizados de **search2Agents.py** and **search2.py** respectivamente. As configurações e usabilidade também foram herdadas do código da Universidade de Berkeley, sendo executadas na seção de Experimentos e discussão e utilizando estas *flags*:

- **-l** - Determina o *layout* de labirinto a ser utilizado
- **-p** - Determina o agente inteligente que controlará o *Pac-Man*
- **-a** - Determina os parâmetros dos algoritmos utilizados. É possível selecionar o algoritmo utilizado (**fn**) e a heurística (**heuristic**)
- **-z** - Zoom. Determina a escala da janela a ser aberta para visualização do jogo (de 0.1 a 0.9)

Os labirintos são gerados em arquivos de texto, em arquivos de extensão **.lay**. Nele, o programa reconhece como códigos os seguintes caracteres:

- **%** - Parede
- **.** - Bolinha
- **o** - Bolinha com poderes
- **G** - Fantasma
- **P** - *Pac-man*

Tal padrão permite um desenvolvimento mais flexível e intuitivo dos labirintos, podendo colocar o *Pac-man*, os fantasmas e as bolinhas em diversas posições e com tamanhos e formatos variáveis de labirintos. Devido aos requisitos do projeto, foram colocadas as seguintes restrições:

- Como o objetivo do problema prioriza a avaliação dos algoritmos, o problema foi modelado com o objetivo mais simples: Encontrar a única bolinha que está presente nos mapas propostos.
- Não foram colocadas bolinhas com poderes nos labirintos. Isso dá mais ênfase a busca, pois o *Pac-man* sempre perderá o jogo se o algoritmo escolher o caminho onde está um dos fantasmas.
- São colocados 3 fantasmas em posições determinadas, porém distintas.
- O agente inteligente dos fantasmas é desligado durante o jogo.
- Para simplificar os *layouts* e testes dos algoritmos, o labirinto possui o formato retangular e fechado.
- O labirinto é feito em **[6]**, colocado em arquivo e possui um tamanho de 28 linhas x 31 colunas
- O *Pac-man* é colocado em uma posição determinada.

### Problema, agentes e ambiente

Como o labirinto é representado por um grid, os estados foram representados por tuplas no formato **(x,y)** e as ações são determinadas pelas direções **['South', 'West', 'East', 'North']**. Uma implementação simples do problema na biblioteca AIMA pode ser verificada no código abaixo:

Nele, a classe **SimplePacmanProblem** é uma classe filha da classe **Problem**, que é passada para os algoritmos de busca encontrarem o melhor caminho. Como parâmetros em seu construtor há um valor inicial, final, obstáculos e heurística. **action_cost** é a função heurística utilizada para algoritmos de busca informada (ou **g(x)**), **h** é a função heurística utilizada para o *A*\* (ou **h(x)**) e **value** é a função heurística utilizada para algoritmos de busca locais (ou **g(x)**). No AIMA, o heurística nos algoritmos de busca locais em seu código original é realizada em direção a um mínimo local. O valor de **directions** seria semelhante às ações que o *Pac-man* executaria no caminho, fazendo estados e ações serem tuplas no formato **(x,y)** e praticamente não tendo diferenciação.

A classe semelhante no código utilizado neste projeto é a **PositionSearchProblem2**, que é uma classe filha da classe **SearchProblem2**, a qual é passada para os algoritmos de busca encontrarem o melhor caminho. Como parâmetros em seu construtor há um valor inicial, final, heurística e o estado do jogo, que guarda as posições das paredes, fantasmas, das bolinhas e do *Pac-man*. **getCostOfActions** é a função utilizada para determinar os estados através da função heurística para os algoritmos de busca informada e local (dependendo dos parâmetros utilizados pode ser **g(x)** ou **h(x)**). **getSuccessors** é a função que disponibiliza os nós sucessores da busca, semelhante a função **actions** presente na AIMA. Há ainda a função **isGoalState**, semelhante a função **goal_test** da classe **Problem**. A classe **Search2Agent** faz o papel de agente inteligente e a ligação entre problema, algoritmo de busca, estado do jogo e heurística.

Um diagrama de classes com todas as classes presentes no código pode ser visto no PDF abaixo:


In [1]:
# This .pdf file shows class diagram that implements how the problem was modeled using game graphical interface. It is part
# of how requirement [R4] and requirement [8] was modeling from Berkeley algorithm design base, and modification done
# to comply with MO416 Project 1 requirements.
import os
cwd = os.getcwd()
cwd=cwd+'/modelling/modellingClasses.pdf'
print(cwd)
import webbrowser
webbrowser.open(r''+os.path.join('modelling','modellingClasses.pdf'))
webbrowser.open(cwd)

C:\Users\Thales E. Nazatto\Documents\Pós\MO416A\MO416AI\Project1/modelling/modellingClasses.pdf


True

## Algoritmos

<center>Tabela de referência dos algoritmos</center>

|Algoritmo|Otimalidade|Completude|Complexidade$^1$ de Tempo|Complexidade$^1$ de Espaço|
|---|---|---|---|---|
|Depth-first Search|Não|Não|$O(b^m)$|$O(bm)$|
|Breadth-first Search|Sim|Sim|$O(b^{d+1})$|$O(b^{d+1})$|
|A*|Sim$^2$|Sim|$O(b^d)$|$O(b^d)$|
|Greedy best-first Search|Não|Depende$^3$|$O(bm)$|$O(bm)$|
|Hill Climbing|Não|Não|$O(b^d)$|$O(1)$|

<center>1. $b$ é número de nós, $d$ é nível, e $m$ é profundidade máxima</center>
<center>2. Minimiza custo quando a heurística é admissível</center>
<center>3. Sim, se não visita nós repetidos. Não, caso contrário.</center>

### Busca não-informada

#### Depth-First Search

In [80]:
# [R1.1] Show dfs (depth first search) method.
!python pacman.py -l easyLayout -p Search2Agent -a fn=dfs -z .6
#!python pacman.py -l normalLayout -p Search2Agent -a fn=dfs -z .6
#!python pacman.py -l hardLayout -p Search2Agent -a fn=dfs -z .6

[Search2Agent] using function dfs
[Search2Agent] using problem type PositionSearchProblem2
[R12] Initial position of pacman is (23, 29)
[R10] Final goal position is (1, 3)
[R11] Ghost Positions is/are [(9, 20), (15, 10), (26, 7)]
Number of foods is 1
[R15] has the game food? False
[R16] Path found with total cost g(x) of 170 in 0.03629422187805176s
[R13] Search nodes expanded: 253
[R13] Nodes visited: [(23, 29), (22, 29), (21, 29), (20, 29), (19, 29), (18, 29), (17, 29), (16, 29), (15, 29), (14, 29), (13, 29), (12, 29), (11, 29), (10, 29), (9, 29), (8, 29), (7, 29), (6, 29), (5, 29), (4, 29), (3, 29), (2, 29), (1, 29), (1, 28), (1, 27), (1, 26), (1, 25), (1, 24), (1, 23), (1, 22), (1, 21), (1, 20), (2, 20), (3, 20), (4, 20), (5, 20), (6, 20), (6, 19), (6, 18), (6, 17), (6, 16), (5, 16), (4, 16), (3, 16), (2, 16), (1, 16), (1, 17), (1, 18), (1, 19), (3, 15), (3, 14), (3, 13), (2, 13), (1, 13), (1, 12), (1, 11), (1, 10), (2, 10), (3, 10), (3, 9), (3, 8), (3, 7), (2, 7), (1, 7), (4, 7), (

#### Breadth-First Search

In [81]:
# [R1.2] Show bfs (breadth first search) method.
!python pacman.py -l easyLayout -p Search2Agent -a fn=bfs -z .6
#!python pacman.py -l normalLayout -p Search2Agent -a fn=bfs -z .6
#!python pacman.py -l hardLayout -p Search2Agent -a fn=bfs -z .6

[Search2Agent] using function bfs
[Search2Agent] using problem type PositionSearchProblem2
[R12] Initial position of pacman is (23, 29)
[R10] Final goal position is (1, 3)
[R11] Ghost Positions is/are [(9, 20), (15, 10), (26, 7)]
Number of foods is 1
[R15] has the game food? False
[R16] Path found with total cost g(x) of 24 in 0.029007434844970703s
[R13] Search nodes expanded: 153
[R13] Nodes visited: [(23, 29), (23, 28), (24, 29), (22, 29), (23, 27), (25, 29), (21, 29), (23, 26), (26, 29), (20, 29), (22, 26), (26, 28), (19, 29), (21, 26), (26, 27), (18, 29), (21, 25), (26, 26), (17, 29), (21, 24), (26, 25), (16, 29), (21, 23), (26, 24), (15, 29), (22, 23), (20, 23), (26, 23), (15, 28), (14, 29), (23, 23), (19, 23), (26, 22), (15, 27), (13, 29), (23, 22), (18, 23), (26, 21), (15, 26), (12, 29), (23, 21), (18, 24), (18, 22), (17, 23), (26, 20), (16, 26), (12, 28), (11, 29), (23, 20), (18, 25), (18, 21), (16, 23), (26, 19), (25, 20), (17, 26), (12, 27), (10, 29), (24, 20), (22, 20), (18,

### Busca informada

#### A*

In [82]:
# [R2.1] Show A* and Manhattan heuristic method.
!python pacman.py -l easyLayout -p Search2Agent -a fn=astar,heuristic=manhattanHeuristic -z .6
#!python pacman.py -l normalLayout -p Search2Agent -a fn=astar,heuristic=manhattanHeuristic -z .6
#!python pacman.py -l hardLayout -p Search2Agent -a fn=astar,heuristic=manhattanHeuristic -z .6

[Search2Agent] using function astar and [R16] heuristic manhattanHeuristic
[Search2Agent] using problem type PositionSearchProblem2
[R12] Initial position of pacman is (23, 29)
[R10] Final goal position is (1, 3)
[R11] Ghost Positions is/are [(9, 20), (15, 10), (26, 7)]
Number of foods is 1
[R15] has the game food? False
[R16] Path found with total cost g(x) of 24 in 0.013302087783813477s
[R13] Search nodes expanded: 49
[R13] Nodes visited: [(23, 29), (23, 28), (22, 29), (23, 27), (21, 29), (23, 26), (20, 29), (22, 26), (19, 29), (21, 26), (18, 29), (21, 25), (17, 29), (21, 24), (16, 29), (15, 29), (15, 28), (14, 29), (15, 27), (13, 29), (15, 26), (12, 29), (12, 28), (11, 29), (12, 27), (10, 29), (12, 26), (9, 29), (11, 26), (8, 29), (10, 26), (7, 29), (9, 26), (6, 29), (9, 25), (9, 24), (24, 29), (21, 23), (16, 26), (5, 29), (9, 23), (20, 23), (8, 23), (19, 23), (7, 23), (18, 23), (6, 23), (18, 24), (17, 23), (6, 24)]
[R13] Solution states: 25 - [(23, 29), (22, 29), (21, 29), (20, 29)

In [83]:
# [R2.1] Show A* and Euclidean heuristic method.
!python pacman.py -l easyLayout -p Search2Agent -a fn=astar,heuristic=euclideanHeuristic -z .6
#!python pacman.py -l normalLayout -p Search2Agent -a fn=astar,heuristic=euclideanHeuristic -z .6
#!python pacman.py -l hardLayout -p Search2Agent -a fn=astar,heuristic=euclideanHeuristic -z .6

[Search2Agent] using function astar and [R16] heuristic euclideanHeuristic
[Search2Agent] using problem type PositionSearchProblem2
[R12] Initial position of pacman is (23, 29)
[R10] Final goal position is (1, 3)
[R11] Ghost Positions is/are [(9, 20), (15, 10), (26, 7)]
Number of foods is 1
[R15] has the game food? False
[R16] Path found with total cost g(x) of 24 in 0.014453649520874023s
[R13] Search nodes expanded: 58
[R13] Nodes visited: [(23, 29), (22, 29), (21, 29), (20, 29), (19, 29), (18, 29), (17, 29), (16, 29), (15, 29), (14, 29), (23, 28), (13, 29), (12, 29), (15, 28), (11, 29), (12, 28), (23, 27), (10, 29), (15, 27), (24, 29), (12, 27), (9, 29), (23, 26), (22, 26), (21, 26), (15, 26), (12, 26), (8, 29), (11, 26), (10, 26), (9, 26), (21, 25), (7, 29), (9, 25), (25, 29), (21, 24), (6, 29), (9, 24), (16, 26), (21, 23), (20, 23), (19, 23), (18, 23), (17, 23), (16, 23), (15, 23), (14, 23), (13, 23), (12, 23), (5, 29), (11, 23), (10, 23), (9, 23), (8, 23), (7, 23), (26, 29), (18, 

#### Greedy Best-First Search

In [85]:
#[R2.2] Show gbfs (greedy best first search) and Manhattan heuristic method.
!python pacman.py -l easyLayout -p Search2Agent -a fn=gbfs,heuristic=manhattanHeuristic -z .6
#!python pacman.py -l normalLayout -p Search2Agent -a fn=gbfs,heuristic=manhattanHeuristic -z .6
#!python pacman.py -l hardLayout -p Search2Agent -a fn=gbfs,heuristic=manhattanHeuristic -z .6

[Search2Agent] using function gbfs and [R16] heuristic manhattanHeuristic
[Search2Agent] using problem type PositionSearchProblem2
[R12] Initial position of pacman is (23, 29)
[R10] Final goal position is (1, 3)
[R11] Ghost Positions is/are [(9, 20), (15, 10), (26, 7)]
Number of foods is 1
[R15] has the game food? False
[R16] Path found with total cost g(x) of 24 in 0.007546901702880859s
[R13] Search nodes expanded: 26
[R13] Nodes visited: [(23, 29), (23, 28), (23, 27), (23, 26), (22, 26), (21, 26), (21, 25), (21, 24), (21, 23), (20, 23), (19, 23), (18, 23), (18, 24), (17, 23), (16, 23), (15, 23), (14, 23), (13, 23), (12, 23), (11, 23), (10, 23), (9, 23), (9, 24), (8, 23), (7, 23), (6, 23), (6, 24)]
[R13] Solution states: 25 - [(23, 29), (23, 28), (23, 27), (23, 26), (22, 26), (21, 26), (21, 25), (21, 24), (21, 23), (20, 23), (19, 23), (18, 23), (17, 23), (16, 23), (15, 23), (14, 23), (13, 23), (12, 23), (11, 23), (10, 23), (9, 23), (8, 23), (7, 23), (6, 23), (6, 24)]
[R14] Solution ac

In [86]:
##[R2.2] Show gbfs (greedy best first search) and Euclidean heuristic method.
!python pacman.py -l easyLayout -p Search2Agent -a fn=gbfs,heuristic=euclideanHeuristic -z .6
#!python pacman.py -l normalLayout -p Search2Agent -a fn=gbfs,heuristic=euclideanHeuristic -z .6
#!python pacman.py -l hardLayout -p Search2Agent -a fn=gbfs,heuristic=euclideanHeuristic -z .6

[Search2Agent] using function gbfs and [R16] heuristic euclideanHeuristic
[Search2Agent] using problem type PositionSearchProblem2
[R12] Initial position of pacman is (23, 29)
[R10] Final goal position is (1, 3)
[R11] Ghost Positions is/are [(9, 20), (15, 10), (26, 7)]
Number of foods is 1
[R15] has the game food? False
[R16] Path found with total cost g(x) of 26 in 0.008487224578857422s
[R13] Search nodes expanded: 26
[R13] Nodes visited: [(23, 29), (22, 29), (21, 29), (20, 29), (19, 29), (18, 29), (17, 29), (16, 29), (15, 29), (14, 29), (13, 29), (12, 29), (11, 29), (10, 29), (9, 29), (8, 29), (7, 29), (6, 29), (5, 29), (4, 29), (4, 28), (4, 27), (4, 26), (5, 26), (6, 26), (6, 25), (6, 24)]
[R13] Solution states: 27 - [(23, 29), (22, 29), (21, 29), (20, 29), (19, 29), (18, 29), (17, 29), (16, 29), (15, 29), (14, 29), (13, 29), (12, 29), (11, 29), (10, 29), (9, 29), (8, 29), (7, 29), (6, 29), (5, 29), (4, 29), (4, 28), (4, 27), (4, 26), (5, 26), (6, 26), (6, 25), (6, 24)]
[R14] Soluti

### Busca local (*Hill Climbing*)

In [118]:
from re import search
from subprocess import run

def run_hcs(layout):
    '''
    Extrai o número de nós visitados e expandidos da
    stdout do comando pacman.py.
    '''
    nodes_visited = 0
    nodes_expanded = 0
    output = run(['python', 'pacman.py', 
                  '-l', layout, 
                  '-p', 'Search2Agent', 
                  '-a', 'fn=hcs', 
                  '-z', '.6'], capture_output=True)
    results = output.stdout.decode('utf-8').split('\n')
    for result in results:
        matchstates = search('Solution states: ', result)
        if matchstates:
            nodes_visited = int(result[matchstates.end():].split(' ')[0])
        matchnodes = search('nodes expanded: ', result)
        if matchnodes:
            nodes_expanded = int(result[matchnodes.end():])
    return nodes_visited, nodes_expanded

def mean_hcs(n, layout):
    '''
    Executa Hill Climbing n vezes no layout escolhido
    e retorna a média de nós visitados e expandidos
    nas n execuções.
    '''
    totalvisited = 0
    totalexpanded = 0
    for _ in range(n):
        nvisited, nexpanded = run_hcs('easyLayout')
        totalvisited += nvisited
        totalexpanded += nexpanded
    meanvisited = totalvisited / n
    meanexpanded = totalexpanded / n
    print(f'{layout}\tNo. nodes visited (mean): {meanvisited}')
    print(f'{layout}t\tNo. nodes expanded (mean): {meanexpanded}')
    return meanvisited, meanexpanded

In [119]:
nvisited, nexpanded = mean_hcs(2, 'easyLayout')

NODES VISITED 55
NODES VISITED 73
easyLayout	No. nodes visited (mean): 64.0
easyLayoutt	No. nodes expanded (mean): 148.0


# Resultados

As tabelas a seguir apresentam o desempenho das implementaçõs.  Estes resultados são comparados com a tabela de referência teórica.

### *easyLayout*

|Algoritmo|Otimalidade|Completude|No. nós expandidos|No. nós visitados|
|---|---|---|---|---|
|Depth-first Search                  |?|Não|253|171|
|Breadth-first Search                |?|Sim|153|25 |
|A* (Euclidean)                      |?|Sim|58 |25 |
|Greedy Best-first Search (Euclidean)|?|Sim|26 |27 |
|A* (Manhattan)                      |?|Sim|49 |25 |
|Greedy Best-first Search (Manhattan)|?|Sim|26 |25 |
|Hill Climbing                       |?|Não|   |   |

### *normalLayout*

|Algoritmo|Otimalidade|Completude|No. nós expandidos|No. nós visitados|
|---|---|---|---|---|
|Depth-first Search                  |?|Não|214|140|
|Breadth-first Search                |?|Sim|203|30 |
|A* (Euclidean)                      |?|Sim|109|30 |
|Greedy Best-first Search (Euclidean)|?|Sim|48 |30 |
|A* (Manhattan)                      |?|Sim|85 |30 |
|Greedy Best-first Search (Manhattan)|?|Sim|31 |32 |
|Hill Climbing                       |?|Não|   |   |

### *hardLayout*

|Algoritmo|Otimalidade|Completude|No. nós expandidos|No. nós visitados|
|---|---|---|---|---|
|Depth-first Search                  |?|Sim|79 |73|
|Breadth-first Search                |?|Sim|336|49|
|A* (Euclidean)                      |?|Sim|263|49|
|Greedy Best-first Search (Euclidean)|?|Sim|72 |49|
|A* (Manhattan)                      |?|Sim|202|49|
|Greedy Best-first Search (Manhattan)|?|Sim|61 |53|
|Hill Climbing                       |?|Não|   |  |

# Conclusão

- Depth-first search não completou nos layouts *easy* e *normal* devido a ordem que os nós são inseridos na pilha. Nestes casos, é inevitável o encontro do Pacman com um Ghost. O caso *hard*, apesar da maior distância, se encontra em um caminho ótimo.
- Breadth-first search e A* expandiram mais nós que os outros algoritmos, como previsto pela complexidade de espaço.
- Greedy Best-first search obteve melhores resultados em todos os casos (com ambas as heurísticas) devido ao layout do labirinto escolhido, que mostrou-se favorável a uma exploração gulosa. 

# Bibliografia

**[1]** Pac-man - Wikipedia. URL: https://en.wikipedia.org/wiki/Pac-Man

**[2]** Project 1. URL: https://drive.google.com/file/d/187lgVekPC0kBmA2AjOm-KkV8Q_lFLeoH/view

**[3]** Search for AIMA 4th edition - Implementation of search algorithms and search problems for AIMA. URLs: https://github.com/aimacode/aima-python/blob/master/search4e.ipynb, https://github.com/aimacode/aima-python/blob/master/search.ipynb

**[4]** UC Berkeley CS188 Intro to AI -- Course Materials. URL: http://ai.berkeley.edu/search.html

**[5]** *tkint* documentation. URL: https://docs.python.org/3/library/tkinter.html

**[6]** Pac-Man Maze Generation. URL: https://shaunlebron.github.io/pacman-mazegen